dropwizard-auth/src/main/java/io/dropwizard/auth/CachingAuthorizer.java

```diff
@@ -0,0 +1,172 @@
+package io.dropwizard.auth;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.Timer;
+import com.google.common.base.Predicate;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheBuilderSpec;
+import com.google.common.cache.CacheStats;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Sets;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+
+import java.security.Principal;
+
+import static com.codahale.metrics.MetricRegistry.name;
+
+/**
+ * An {@link Authorizer} decorator which uses a Guava {@link Cache} to
+ * temporarily cache principals' role associations.
+ *
+ * Cache entries include both inclusion and exclusion of a principal
+ * within a given role.
+ *
+ * @param <P> the type of principals on which the authorizer operates
+ */
+public class CachingAuthorizer<P extends Principal> implements Authorizer<P> {
+    private final Authorizer<P> underlying;
+    private final Cache<ImmutablePair<P, String>, Boolean> cache;
```

Comment:
what about the value being an enum:
```java
private class AuthzResult {
  Allow, Deny
}
```
so it's rather clearer that the tri-state of `true`, `false`, `null` in the boolean case.